app/src/ui/check-runs/ci-check-run-list.tsx

```diff
@@ -52,30 +53,46 @@ export class CICheckRunList extends React.PureComponent<
   }
 
   public componentDidUpdate(prevProps: ICICheckRunListProps) {
-    this.setState(
-      this.setupStateAfterCheckRunPropChange(this.props, this.state)
-    )
+    const {
+      checkRunExpanded,
+      hasUserToggledCheckRun,
+    } = this.setupStateAfterCheckRunPropChange(this.props, this.state)
+    this.setState({ checkRunExpanded, hasUserToggledCheckRun })
   }
 
   private setupStateAfterCheckRunPropChange(
     props: ICICheckRunListProps,
     currentState: ICICheckRunListState | null
   ): ICICheckRunListState {
+    // If the user has expanded something and then a load occurs, we don't want
+    // to reset their position.
+    if (currentState !== null && currentState.hasUserToggledCheckRun) {
+      return currentState
+    }
+
+    const checkRunGroups =
+      currentState === null
+        ? getCheckRunsGroupedByActionWorkflowNameAndEvent(props.checkRuns)
+        : currentState.checkRunGroups
+
     let checkRunExpanded = currentState?.checkRunExpanded ?? null
 
-    if (currentState === null || !currentState.hasUserToggledCheckRun) {
-      // If there is a failure, we want the first check run with a failure, to
-      // be opened so the user doesn't have to click through to find it.
-      // Otherwise, just open the first one. (Only actions type can be expanded.)
-      const firstFailure = props.checkRuns.find(
+    // If there is a failure, we want the first check run with a failure, to
+    // be opened so the user doesn't have to click through to find it.
+    let firstFailure: IRefCheck | undefined
+    checkRunExpanded = null
+    for (const group of checkRunGroups.values()) {
+      firstFailure = group.find(
```

Comment:
Not sure if I'm missing anything, but is there a reason to move this out of the `for`? AFAICT `firstFailure` is not being changed outside of the `for` so it can still be `const`, right? 🤔

Reply:
Yep - Sorry left over from old approach.